lib/rules/func-name-matching.js

```diff
@@ -0,0 +1,151 @@
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "require function names to match the name of the variable or property to which they are assigned",
+            category: "Stylistic Issues",
+            recommended: false
+        },
+
+        schema: [
+            {
+                type: "object",
+                properties: {
+                    includeCommonJSModuleExports: {
+                        type: "boolean"
+                    }
+                },
+                additionalProperties: false
+            }
+        ]
+    },
+
+    create(context) {
+
+        const includeModuleExports = context.options[0] && context.options[0].includeCommonJSModuleExports;
+
+        /**
+         * Reports
+         * @param {ASTNode} node The node to report
+         * @param {string} name The variable or property name
+         * @param {string} funcName The function name
+         * @param {boolean} isProp True if the reported node is a property assignment
+         * @returns {void}
+         */
+        function report(node, name, funcName, isProp) {
+            context.report({
+                node,
+                message: isProp ? "Function name `{{funcName}}` should match property name `{{name}}`"
+                                : "Function name `{{funcName}}` should match variable name `{{name}}`",
+                data: {
+                    name,
+                    funcName
+                }
+            });
+        }
+
+        //--------------------------------------------------------------------------
+        // Public
+        //--------------------------------------------------------------------------
+
+        return {
+
+            VariableDeclarator(node) {
+                if (node.init.type !== "FunctionExpression") {
+                    return;
+                }
+                if (node.init.id && node.id.name !== node.init.id.name) {
+                    report(node, node.id.name, node.init.id.name, false);
+                }
+            },
+
+            AssignmentExpression(node) {
+                if (node.right.type !== "FunctionExpression" || (!includeModuleExports && isModuleExports(node.left))) {
+                    return;
+                }
+                const name = getName(node.left);
+                const isProp = node.left.type === "MemberExpression" ? true : false;
+
+                if (node.right.id && name.length > 0 && name !== node.right.id.name) {
+                    report(node, name, node.right.id.name, isProp);
+                }
+            },
+
+            Property(node) {
+                if (node.value.type !== "FunctionExpression" || node.key.type !== "Identifier") {
+                    return;
+                }
+                if (node.value.id && node.key.name !== node.value.id.name) {
```

Comment:
I think this logic needs a few tweak.

1. For `LiteralPropertyName`s: `({"foo": function bar() {}})`;  
   In this case, `node.key.type` is `"Literal"`.
2. For Computed Properties: `({[foo]: function bar() {}})`  
   In `AssignmentExpression`s cases, it seems to ignore `Identifier`s in computed properties. I love consistency.
3. `Literal` might not convert to `Identifier`. I think this rule should ignore such literals.  
  `({"foo//bar": function bar() {}})`

Maybe `astUtils.getStaticPropertyName(node)` can be used to get the name of properties.